src/Sensor/SensorBms.cpp

```diff
@@ -0,0 +1,144 @@
+            CanMessage msg = CAN_MESSAGE_NULL;
+            msg.id = CAN_BMS_REQUEST;
+            msg.dataLength = REQ_DATA_LENGTH;
+            msg.data[0] = _paramIds[i];
+            
+            _canInterface.sendMessage(msg);
+        }
+        
+        _lastValidTime = millis();
+    }
+}
+
+String SensorBms::getHumanName() {
+    return "BMS";
+}
+
+void SensorBms::update(CanMessage message) {
+    DEBUG_SERIAL_LN("RECEIVED RESPONSE FROM BMS");
+
+    if(message.data[RSP_STATUS_BYTE] != TRUE) {
+        DEBUG_SERIAL_LN("Poor Data Received");
+    }
+    else {
+        unsigned statusCode = 0;
+        switch (message.data[RSP_PARAM_ID_BYTE]) {
+            case PARAM_ID_BATTERY_VOLTAGE:
+                _batteryVoltage = parseFloat(message.data + RSP_DATA_BYTE);
+                break;
+            case PARAM_ID_BATTERY_CURRENT:
+                _batteryCurrent = parseFloat(message.data + RSP_DATA_BYTE);
+                break;
+            case PARAM_ID_MAX_CELL_VOLTAGE:
+                _cellVoltageMax = parseFloat(message.data + RSP_DATA_BYTE);    
+                break;
+            case PARAM_ID_MIN_CELL_VOLTAGE:
+                _cellVoltageMin = parseFloat(message.data + RSP_DATA_BYTE);    
+                break;
+            case PARAM_ID_STATUS:
+                statusCode = (message.data[3] << 8) + message.data[2];
+                if(statusCode == STATUS_CHARGING) {
+                    _bmsStatus = "BMS Charging...";
+                    DEBUG_SERIAL_LN("BMS Charging...");
+                }
+                else if(statusCode == STATUS_CHARGED) {
+                    _bmsStatus = "BMS Charged!";
+                    DEBUG_SERIAL_LN("BMS Charged!");
+                }
+                else if(statusCode == STATUS_DISCHARGING) {
+                    _bmsStatus = "BMS Discharging...";
+                    DEBUG_SERIAL_LN("BMS Discharging...");
+                }
+                else if(statusCode == STATUS_REGENERATION) {
+                    _bmsStatus = "BMS Regeneration";
+                    DEBUG_SERIAL_LN("BMS Regeneration");
+                }
+                else if(statusCode == STATUS_IDLE) {
+                    _bmsStatus = "BMS Idle";
+                    DEBUG_SERIAL_LN("BMS Idle");
+                }
+                else {
+                    _bmsStatus = "BMS Fault Error";
+                    DEBUG_SERIAL_LN("BMS Fault Error");
+                }
+                break;
+            case PARAM_ID_SOC:
+                _soc = parseFloat(message.data + RSP_DATA_BYTE);    
+                break;
+            case PARAM_ID_TEMP:
+                if(message.data[5] == TEMP_ID_INTERNAL) {
+                    _tempBms = (message.data[4] << 8) + message.data[3];
+                }
+                else if(message.data[5] == TEMP_ID_BATTERY_1) {
+                    _batteryTemp1 = (message.data[4] << 8) + message.data[3];
+                }
+                else {
```

Comment:
I think this should also be an "else if" statement checking for TEMP_ID_BATTERY_2, just in case the BMS randomly sends out another ID. If it does, we wouldn't want to record it at all. I'm thinking in the case we update the BMS firmware and it adds another temp sensor or something.

Reply:
Should I include a serial message for that empty else case or just not include it?

Reply:
Yeah, I think it's ok without a serial debug message